<!DOCTYPE html>
<html lang="en" class="js csstransforms3d">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta name="generator" content="Hugo 0.80.0" />
    <meta name="description" content="Single-cell RNA-seq analysis with R/Bioconductor">
<meta name="author" content="Jacques Serizay">

    <link rel="icon" href="/scRNAseq_Physalia_2022/images/favicon.png" type="image/png">

    <title>2. Program :: Single-cell RNA-seq analysis with R/Bioconductor</title>

    
    <link href="/scRNAseq_Physalia_2022/css/nucleus.css?1654458090" rel="stylesheet">
    <link href="/scRNAseq_Physalia_2022/css/fontawesome-all.min.css?1654458090" rel="stylesheet">
    <link href="/scRNAseq_Physalia_2022/css/hybrid.css?1654458090" rel="stylesheet">
    <link href="/scRNAseq_Physalia_2022/css/featherlight.min.css?1654458090" rel="stylesheet">
    <link href="/scRNAseq_Physalia_2022/css/perfect-scrollbar.min.css?1654458090" rel="stylesheet">
    <link href="/scRNAseq_Physalia_2022/css/auto-complete.css?1654458090" rel="stylesheet">
    <link href="/scRNAseq_Physalia_2022/css/atom-one-dark-reasonable.css?1654458090" rel="stylesheet">
    <link href="/scRNAseq_Physalia_2022/css/theme.css?1654458090" rel="stylesheet">
    <link href="/scRNAseq_Physalia_2022/css/hugo-theme.css?1654458090" rel="stylesheet">
    
    <link href="/scRNAseq_Physalia_2022/css/theme-red.css?1654458090" rel="stylesheet">
    
    

    <script src="/scRNAseq_Physalia_2022/js/jquery-3.3.1.min.js?1654458090"></script>

    <style>
      :root #header + #content > #left > #rlblock_left{
          display:none !important;
      }
      
    </style>
    
  </head>
  <body class="" data-url="/scRNAseq_Physalia_2022/introduction/02_program/">
    <nav id="sidebar" class="">



  <div id="header-wrapper">
    <div id="header">
      <a id="logo">



</a>

    </div>
    
  </div>
  

    <div class="highlightable">
    <ul class="topics">

        
          
          




 
  
    
    <li data-nav-id="/scRNAseq_Physalia_2022/introduction/" title=" Single-cell RNA-seq analysis with R/Bioconductor" class="dd-item
        parent
        
        
        ">
      <a href="/scRNAseq_Physalia_2022/introduction/">
           Single-cell RNA-seq analysis with R/Bioconductor
          
      </a>
      
      
        <ul>
          
          
          

        
          
            
            




 
  
    
      <li data-nav-id="/scRNAseq_Physalia_2022/introduction/01_presentation/" title="1. Presentation" class="dd-item ">
        <a href="/scRNAseq_Physalia_2022/introduction/01_presentation/">
        1. Presentation
        
        </a>
    </li>
     
  
 

            
          
            
            




 
  
    
      <li data-nav-id="/scRNAseq_Physalia_2022/introduction/02_program/" title="2. Program" class="dd-item active">
        <a href="/scRNAseq_Physalia_2022/introduction/02_program/">
        2. Program
        
        </a>
    </li>
     
  
 

            
          
            
            




 
  
    
      <li data-nav-id="/scRNAseq_Physalia_2022/introduction/03_contacts/" title="3. Contacts" class="dd-item ">
        <a href="/scRNAseq_Physalia_2022/introduction/03_contacts/">
        3. Contacts
        
        </a>
    </li>
     
  
 

            
          
            
            




 
  
    
      <li data-nav-id="/scRNAseq_Physalia_2022/introduction/04_rstudio/" title="4. Rstudio" class="dd-item ">
        <a href="/scRNAseq_Physalia_2022/introduction/04_rstudio/">
        4. Rstudio
        
        </a>
    </li>
     
  
 

            
          
            
            




 
  
    
      <li data-nav-id="/scRNAseq_Physalia_2022/introduction/05_prerequisites/" title="5. Prerequisites and required config" class="dd-item ">
        <a href="/scRNAseq_Physalia_2022/introduction/05_prerequisites/">
        5. Prerequisites and required config
        
        </a>
    </li>
     
  
 

            
          
        
        </ul>
      
    </li>
  
 

          
          




 
  
    
    <li data-nav-id="/scRNAseq_Physalia_2022/day01/" title="1. Day 01" class="dd-item
        
        
        
        ">
      <a href="/scRNAseq_Physalia_2022/day01/">
          1. Day 01
          
      </a>
      
      
    </li>
  
 

          
          




 
  
    
    <li data-nav-id="/scRNAseq_Physalia_2022/day02/" title="2. Day 02" class="dd-item
        
        
        
        ">
      <a href="/scRNAseq_Physalia_2022/day02/">
          2. Day 02
          
      </a>
      
      
    </li>
  
 

          
          




 
  
    
    <li data-nav-id="/scRNAseq_Physalia_2022/day03/" title="3. Day 03" class="dd-item
        
        
        
        ">
      <a href="/scRNAseq_Physalia_2022/day03/">
          3. Day 03
          
      </a>
      
      
    </li>
  
 

          
          




 
  
    
    <li data-nav-id="/scRNAseq_Physalia_2022/day04/" title="4. Day 04" class="dd-item
        
        
        
        ">
      <a href="/scRNAseq_Physalia_2022/day04/">
          4. Day 04
          
      </a>
      
      
    </li>
  
 

          
          




 
  
    
    <li data-nav-id="/scRNAseq_Physalia_2022/day05/" title="5. Day 05" class="dd-item
        
        
        
        ">
      <a href="/scRNAseq_Physalia_2022/day05/">
          5. Day 05
          
      </a>
      
      
    </li>
  
 

          
          




 
  
    
    <li data-nav-id="/scRNAseq_Physalia_2022/extra-resources/" title="Extra resources" class="dd-item
        
        
        
        ">
      <a href="/scRNAseq_Physalia_2022/extra-resources/">
          Extra resources
          
      </a>
      
      
    </li>
  
 

          
        
    </ul>

    
    

    
    <section id="footer">
      <center>

    <a href="https://github.com/js2264/scRNAseq_Physalia_2021">Check this website source on Github</a>
    <p>Created by J. Serizay, O. Ashenberg & A. Mehta</p>
    <p>Copyright 2022</p>

</center>

<script async defer src="https://buttons.github.io/buttons.js"></script>

    </section>
  </div>
</nav>




        <section id="body">
        <div id="overlay"></div>
        <div class="padding highlightable">
        <div id="head-tags">
        
        </div>
        
        <div id="body-inner">
          
            <h1>
              
              2. Program
            </h1>
          

        


<p>Classes are from:</p>
<ul>
<li>2 to 8 pm Paris time.</li>
<li>1 to 7 pm London time.</li>
<li>8 am to 2 pm NY time.</li>
<li>5 am to 11 am SF time.</li>
</ul>
<h2 id="monday--classes-from-1400-to-2000-paris-time">Monday – Classes from 14:00 to 20:00 (Paris time)</h2>
<h3 id="lecture-1--introduction-to-scrna-seq-analysis-orr">Lecture 1 – Introduction to scRNA-Seq analysis [Orr]</h3>
<ul>
<li>General introduction: cell atlas overviews</li>
<li>Comparison of bulk and single cell RNA-Seq</li>
<li>Overview of available scRNA-seq technologies (10x) and experimental protocols</li>
</ul>
<h3 id="lab-1--familiarizing-yourself-with-the-course-aws-instance-jacques">Lab 1 – Familiarizing yourself with the course AWS instance [Jacques]</h3>
<ul>
<li>Using RStudio</li>
<li>Logging in AWS</li>
<li>Shell and Unix commands to navigate directories, create folders, open files</li>
<li>Raw file formats</li>
<li>Get data from 10x website, single cell portal, from GEO (fastqs, counts)</li>
</ul>
<h3 id="lecture-2---from-sequencing-reads-to-expression-matrices-jacques">Lecture 2 - From sequencing reads to expression matrices [Jacques]</h3>
<ul>
<li>scRNA-Seq processing workflow starting with choice of sequencer (NextSeq, HiSeq, MiSeq) / barcode swapping and bcl files</li>
<li>Overview of Popular tools and algorithms</li>
<li>Common single-cell analyses and interpretation</li>
<li>Sequencing data: alignment and quality control</li>
<li>IGV: Looking at cool things in alignment like where reads are, mutations, splicing</li>
</ul>
<h3 id="lab-2--from-sequencing-reads-to-expression-matrices-arnav">Lab 2 – From sequencing reads to expression matrices [Arnav]</h3>
<ul>
<li>Mapping sequencing data with Cellranger</li>
<li>Quality Control reports (CellRanger, dropEst, fastqc)</li>
</ul>
<h3 id="flash-talks-everybody">Flash talks [Everybody]</h3>
<h2 id="tuesday--classes-from-1400-to-2000-paris-time">Tuesday – Classes from 14:00 to 20:00 (Paris time)</h2>
<h3 id="lab-3---introduction-to-rbioconductor-jacques">Lab 3 - Introduction to R/Bioconductor [Jacques]</h3>
<ul>
<li>Installing packages with CRAN and Bioconductor</li>
<li>Data types, data manipulation, slicing</li>
<li>I/O for scRNAseq analysis in R</li>
</ul>
<h3 id="lecture-3---quality-control-for-scrna-seq-data-arnav">Lecture 3 - Quality control for scRNA-Seq data [Arnav]</h3>
<ul>
<li>What CellRanger does for quality filtering</li>
<li>Normalisation methods <a href="https://www.nature.com/articles/nmeth.4292">https://www.nature.com/articles/nmeth.4292</a></li>
<li>Doublets, empty droplets, DropletUtils</li>
<li>Barcode swapping</li>
<li>Regression with technical covariates</li>
</ul>
<h3 id="lecture-4---identifying-cell-populations-jacques">Lecture 4 - Identifying cell populations [Jacques]</h3>
<ul>
<li>Feature selection</li>
<li>Dimensionality reduction</li>
<li>Graph-based clustering and other cluster methods</li>
<li>Assigning cluster identity</li>
<li>Differential expression tests</li>
</ul>
<h3 id="lab-4--scrna-seq-data-wrangling-orr">Lab 4 – scRNA-Seq data wrangling [Orr]</h3>
<ul>
<li>Data structure</li>
<li>Data filtering</li>
<li>Exploratory data analysis</li>
</ul>
<h2 id="wednesday--classes-from-1400-to-2000-paris-time">Wednesday – Classes from 14:00 to 20:00 (Paris time)</h2>
<h3 id="lab-5--identifying-cell-populations-dimensionality-reduction-clustering-and-annotation-jacques">Lab 5 – Identifying Cell Populations: dimensionality reduction, clustering and annotation [Jacques]</h3>
<ul>
<li>Feature selection</li>
<li>Dimensional reduction</li>
<li>Graph-based clustering</li>
<li>Marker gene detection</li>
<li>Cell type annotation</li>
<li>Data visualization</li>
</ul>
<h3 id="lecture-5---data-integration-and-batch-effect-correction-orr">Lecture 5 - Data integration and batch effect correction [Orr]</h3>
<ul>
<li>Batch correction methods (regress out batch, scaling within batch, Seurat v3, MNN, Liger, Harmony, scvi, scgen)</li>
<li>Evaluation methods for batch correction (ARI, average silhouette width, kBET…)</li>
</ul>
<h3 id="lab-6---data-integration-and-batch-effect-correction-orr">Lab 6 - Data integration and batch effect correction [Orr]</h3>
<ul>
<li>Comparison of batch correction methods</li>
<li>Choosing the optimal batch correction approach</li>
</ul>
<h3 id="lecture-6---trajectories-and-pseudotimes-jacques">Lecture 6 - Trajectories and pseudotimes [Jacques]</h3>
<ul>
<li>Trajectory inference</li>
<li>Popular tools and packages for trajectory analysis (<a href="https://github.com/dynverse/dynmethods#list-of-included-methods">https://github.com/dynverse/dynmethods#list-of-included-methods</a>)</li>
<li>Pseudotime inference</li>
<li>RNA velocity</li>
<li>Differential expression through pseudotime</li>
</ul>
<h2 id="thursday--classes-from-1400-to-2000-paris-time">Thursday – Classes from 14:00 to 20:00 (Paris time)</h2>
<h3 id="lab-7---inferring-differentiation-trajectories-and-pseudotime-jacques">Lab 7 - Inferring differentiation trajectories and pseudotime [Jacques]</h3>
<ul>
<li>Infering trajectory in sperm cell lineage</li>
<li>Orientating a trajectory with RNA veloctiy</li>
<li>DE analysis along a trajectory</li>
</ul>
<h3 id="lecture-7---advances-in-single-cell-genomics-the-epigenome-arnav">Lecture 7 - Advances in single-cell genomics: the epigenome [Arnav]</h3>
<h3 id="lecture-8---advances-in-single-cell-genomics-spatial-transcriptomics-orr">Lecture 8 - Advances in single-cell genomics: spatial transcriptomics [Orr]</h3>
<h3 id="lab-8---single-cell-atac-seq-analysis-arnav">Lab 8 - Single-cell ATAC-Seq analysis [Arnav]</h3>
<h2 id="friday--classes-from-1400-to-2000-paris-time">Friday – Classes from 14:00 to 20:00 (Paris time)</h2>
<p>Friday will be divided in three parts:</p>
<ul>
<li>Morning (1h): &ldquo;Journal club&rdquo;: discussion of method papers</li>
<li>Morning &amp; afternoon (1h + 1h30): Group projects: analysing scRNA-seq data by yourself, from A to Z</li>
<li>Afternoon (1h): Group presentations (10' each group, max 5 slides: what/why/where/when/how, conclusions)</li>
</ul>
<p>Happy hour time!!</p>


<footer class="footline">
	
</footer>

        
        </div>
        

      </div>

    <div id="navigation">
        
        

        
            
            
                
                    
                    
                
                

                    
                    
                        
                    
                    

                    
                        
            
            
                
                    
                    
                
                

                    
                    
                    

                    
                        
            
            
                
                    
                        
                        
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                        
                        
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
            
        

        


	 
	 
    </div>

    </section>

    <div style="left: -1000px; overflow: scroll; position: absolute; top: -1000px; border: none; box-sizing: content-box; height: 200px; margin: 0px; padding: 0px; width: 200px;">
      <div style="border: none; box-sizing: content-box; height: 200px; margin: 0px; padding: 0px; width: 200px;"></div>
    </div>
    <script src="/scRNAseq_Physalia_2022/js/clipboard.min.js?1654458090"></script>
    <script src="/scRNAseq_Physalia_2022/js/perfect-scrollbar.min.js?1654458090"></script>
    <script src="/scRNAseq_Physalia_2022/js/perfect-scrollbar.jquery.min.js?1654458090"></script>
    <script src="/scRNAseq_Physalia_2022/js/jquery.sticky.js?1654458090"></script>
    <script src="/scRNAseq_Physalia_2022/js/featherlight.min.js?1654458090"></script>
    <script src="/scRNAseq_Physalia_2022/js/highlight.pack.js?1654458090"></script>
    <script>hljs.initHighlightingOnLoad();</script>
    <script src="/scRNAseq_Physalia_2022/js/modernizr.custom-3.6.0.js?1654458090"></script>
    <script src="/scRNAseq_Physalia_2022/js/learn.js?1654458090"></script>
    <script src="/scRNAseq_Physalia_2022/js/hugo-learn.js?1654458090"></script>
    
        
            <script src="/scRNAseq_Physalia_2022/mermaid/mermaid.js?1654458090"></script>
        
        <script>
            mermaid.initialize({ startOnLoad: true });
        </script>
    
    <script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-105947713-1', 'auto');
  ga('send', 'pageview');

</script>

  </body>
</html>

